di-ipv-credential-issuer-stub/src/main/java/uk/gov/di/ipv/stub/cred/service/AuthCodeService.java

```diff
@@ -1,23 +1,38 @@
 package uk.gov.di.ipv.stub.cred.service;
 
 import com.nimbusds.oauth2.sdk.AuthorizationCode;
+import uk.gov.di.ipv.stub.cred.error.CriStubException;
 
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 
 public class AuthCodeService {
-    private final Map<String, String> authCodes;
+    private static final String RESOURCE_PAYLOAD = "resourcePayload";
+    public static final String REDIRECT_URL = "redirectUrl";
+    private final Map<String, Map<String, String>> authCodes;
 
     public AuthCodeService() {
         this.authCodes = new ConcurrentHashMap<>();
     }
 
-    public void persist(AuthorizationCode authCode, String resourcePayload) {
-        this.authCodes.put(authCode.getValue(), resourcePayload);
+    public void persist(AuthorizationCode authCode, String resourcePayload, String redirectUrl) {
+        this.authCodes.put(authCode.getValue(), Map.of(RESOURCE_PAYLOAD, resourcePayload, REDIRECT_URL, redirectUrl));
     }
 
     public String getPayload(String authCode) {
-        return this.authCodes.get(authCode);
+        try {
+            return this.authCodes.get(authCode).get(RESOURCE_PAYLOAD);
+        } catch (NullPointerException e) {
```

Comment:
Generally I'd avoid trying to catch NullPointerException instead of testing against null values before using it. But I guess we don't need to be as strict with this being a Stub so fine to leave as is but might be a better alternative to test this.authCodes.get(authCode) is null or not first before trying to get the resource payload?

Reply:
Yep, fair. I'll test for null then return.